yjit/src/codegen.rs

```diff
@@ -5893,6 +5893,95 @@ fn jit_rb_str_empty_p(
     return true;
 }
 
+// Codegen for rb_str_concat() with an integer argument -- *not* String#concat
+// Using strings as a byte buffer often includes appending byte values to the end of the string.
+fn jit_rb_str_concat_codepoint(
+    jit: &mut JITState,
+    asm: &mut Assembler,
+    ocb: &mut OutlinedCb,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<crate::codegen::BlockHandler>,
+    _argc: i32,
+    _known_recv_class: Option<VALUE>,
+) -> bool {
+    asm_comment!(asm, "String#<< with codepoint argument");
+
+    // Ensure the codepoint argument is a Fixnum.
+    let arg = asm.stack_opnd(0);
+    let comptime_arg = jit.peek_at_stack(&asm.ctx, 0);
+    if comptime_arg.fixnum_p() {
+        jit_guard_known_klass(
+            jit,
+            asm,
+            ocb,
+            comptime_arg.class_of(),
+            arg,
+            arg.into(),
+            comptime_arg,
+            SEND_MAX_DEPTH,
+            Counter::guard_send_not_fixnums,
+        );
+    } else if ! unsafe { RB_TYPE_P(comptime_arg, RUBY_T_STRING) } {
+        return jit_rb_str_concat(jit, asm, ocb, _ci, _cme, _block, _argc, _known_recv_class);
+    } else {
```

Comment:
Feels a bit weird to see this check here since you already checked in `jit_rb_str_concat`. You could just go straight to return false.

Reply:
Thanks. I was hoping someone would weigh in on this one. I wasn't sure to what extent we can make assumptions given we know all the current callers vs needing to be safe in the presence of a new caller. I'll remove this.